src/main/java/io/bastillion/manage/util/ExternalAuthUtil.java

```diff
@@ -84,175 +78,108 @@
-                                    user.setFirstNm(name[0]);
-                                    user.setLastNm(name[name.length - 1]);
-                                }
-                            }
-
-                            //set email
-                            if (StringUtils.isNotEmpty(email)) {
-                                user.setEmail(email);
-                            } else if (auth.getUsername().contains("@")) {
-                                user.setEmail(auth.getUsername());
-                            }
-                            user.setId(UserDB.insertUser(con, user));
-                        }
-
-                        //assign profiles for user
-                        UserProfileDB.assignProfilesToUser(con, user.getId(), allRoles, allRoles);
-
-                        dirContext.close();
-                        loginModule.commit();
-
-                    } else {
-
-                        Subject subject = loginContext.getSubject();
-
-                        if (user == null) {
-                            user = new User();
-
-                            user.setUserType(User.ADMINISTRATOR);
-                            user.setUsername(auth.getUsername());
-
-                            //if it looks like name is returned default it
-                            for (Principal p : subject.getPrincipals()) {
-                                if (p.getName().contains(" ")) {
-                                    String[] name = p.getName().split(" ");
-                                    if (name.length > 1) {
-                                        user.setFirstNm(name[0]);
-                                        user.setLastNm(name[name.length - 1]);
-                                    }
-                                }
-                            }
-
-                            //set email
-                            if (auth.getUsername().contains("@")) {
-                                user.setEmail(auth.getUsername());
-                            }
-                            user.setId(UserDB.insertUser(con, user));
-                        }
-
-                    }
-                    authToken = UUID.randomUUID().toString();
-                    user.setAuthToken(authToken);
-                    user.setAuthType(Auth.AUTH_EXTERNAL);
-                    //set auth token
-                    AuthDB.updateLogin(con, user);
+                if (user == null) {
+                    user = createUser(subject);
+                    user.setId(UserDB.insertUser(con, user));
                 }
+                authToken = UUID.randomUUID().toString();
+                user.setAuthToken(authToken);
+                user.setAuthType(Auth.AUTH_EXTERNAL);
+                //set auth token
+                AuthDB.updateLogin(con, user);
+                List<String> userRoles = getUserRoles(subject);
+                UserProfileDB.assignProfilesToUser(con, user.getId(), userRoles, userRoles);
```

Comment:
I think this assignProfilesToUser needs refactored. I don't understand why we need two userRoles params.

Reply:
I agree. Although this method is just called once I didn't want to touch it yet. In the original code, allProfilesNmList was passed a list of all the roleNameAttribute values found under the roleBaseDn subtree and only tried to insert users for the profiles which were also in the assignedProfilesNmList. Since profiles must exist in the database it didn't made much sense to me so I passed the same list to both parameters to have the desired effect. I didn't refactored it yet because maybe in some situation it could be useful to pass a list of "permitted" profiles.

Reply:
Added another assignProfilesToUser method which has only one profile list argument. The previous method is not used and could be removed if not useful anymore.